client/client.go

```diff
@@ -22,70 +18,48 @@ import (
-	pedersenECCommitter *commitments.PedersenECCommitter
-	schnorrProver       *dlogproofs.SchnorrProver
-	schnorrECProver     *dlogproofs.SchnorrECProver
-	paillierEncryptor   *encryption.CSPaillier
+type GenericClient struct {
+	id       int32
+	conn     *grpc.ClientConn
+	pbClient pb.ProtocolClient
+	stream   pb.Protocol_RunClient
 }
 
-// ProtocolParams is a map containing values required to bootstrap
-// the chosen protocol.
-type ProtocolParams map[string]*big.Int
-
-func getConnection(serverEndpoint string) (*grpc.ClientConn, error) {
-	logger.Debug("Getting the connection")
-	timeoutSec := config.LoadTimeout()
-	dialOptions := []grpc.DialOption{
-		grpc.WithInsecure(),
-		grpc.WithBlock(),
-		grpc.WithTimeout(time.Duration(timeoutSec) * time.Second),
-	}
-	conn, err := grpc.Dial(serverEndpoint, dialOptions...)
-	if err != nil {
-		return nil, fmt.Errorf("Could not connect to server %v (%v)", serverEndpoint, err)
+func (c *GenericClient) send(msg *pb.Message) error {
+	if err := c.stream.Send(msg); err != nil {
+		return fmt.Errorf("[Client %v] Error sending message: %v", c.id, err)
 	}
-	return conn, nil
+	logger.Infof("[Client %v] Successfully sent request:", c.id, msg)
+
+	return nil
 }
 
-func getStream(client pb.ProtocolClient) (pb.Protocol_RunClient, error) {
-	logger.Debug("Getting the stream")
-	stream, err := client.Run(context.Background())
-	if err != nil {
-		return nil, fmt.Errorf("Error creating the stream: %v", err)
+func (c *GenericClient) receive() (*pb.Message, error) {
+	resp, err := c.stream.Recv()
+	if err == io.EOF {
+		return nil, fmt.Errorf("[Client %v] EOF error", c.id)
+	} else if err != nil {
+		return nil, fmt.Errorf("[Client %v] An error ocurred: %v", c.id, err)
 	}
-	return stream, nil
+	logger.Infof("[Client %v] Received response from the stream: %v", c.id, resp)
+	return resp, nil
 }
 
-func validateSchema(schemaType pb.SchemaType, schemaVariant pb.SchemaVariant) error {
-	if _, success := pb.SchemaType_name[int32(schemaType)]; !success {
-		return fmt.Errorf("Invalid SchemaType: %v", schemaType)
+// close closes the communication stream and connection to the server.
+func (c *GenericClient) close() error {
+	if err := c.stream.CloseSend(); err != nil {
+		return fmt.Errorf("[Client %v] Error closing stream: %v", c.id, err)
 	}
-	if _, success := pb.SchemaVariant_name[int32(schemaVariant)]; !success {
-		return fmt.Errorf("Invalid SchemaVariant: %v", schemaVariant)
+	if err := c.conn.Close(); err != nil {
+		return fmt.Errorf("[Client %v] Error closing connection: %v", c.id, err)
 	}
 	return nil
 }
 
-func NewProtocolClient(endpoint string, schemaType pb.SchemaType,
-	schemaVariant pb.SchemaVariant) (*Client, error) {
-	logger.Debugf("Creating client [SchemaType = %v][SchemaVariant = %v]", schemaType,
-		schemaVariant)
-
-	if err := validateSchema(schemaType, schemaVariant); err != nil {
-		return nil, err
-	}
-
+func NewGenericClient(endpoint string) (*GenericClient, error) {
```

Comment:
Actually, I am using this function to initialize the embedded struct. I call it when I create any protocol-specific client, for instance in `client/cspaillier.go` we have
```go
 func NewCSPaillierClient(endpoint string, pubKeyPath string, m, l *big.Int) (*CSPaillierClient, error) {
   	genericClient, err := NewGenericClient(endpoint)
    	if err != nil {
    		return nil, err
       }
    // ... initialize CSPaillierClient
}
``` 
It's similar in other protocol clients.